File: 백준/Gold/15926. 현욱은 괄호왕이야！！/현욱은 괄호왕이야！！.cc
#include <bits/stdc++.h>

using namespace std;

int n, cnt, ret;
string s;
stack<int> stk;
int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    cin >> n >> s;
    stk.push(-1);
    for(int i = 0; i < n; i++) {
        if(s[i] == '(') {
            stk.push(i);
        }
        if(s[i] == ')') {
            stk.pop();
            if(!stk.empty()) {
                ret = max(ret, i - stk.top());
            } else {
                stk.push(i);
            }
        }
    }
    cout << ret << '\n';
    return 0;
}

File: 백준/Gold/15926. 현욱은 괄호왕이야！！/현욱은 괄호왕이야！！.swift
import Foundation

func main() {
    let n = Int(readLine()!)!
    
    let str = Array(readLine()!)
    var stk = [Int]()
    var d = [Int](repeating: 0, count: n + 1)
    
    for i in 0..<str.count {
        if str[i] == "(" { stk.append(i) }
        else if !stk.isEmpty {
            d[stk.last!] = 1
            d[i] = 1
            _ = stk.popLast()
        }
    }
    
    var cnt = 0
    var ret = 0
    
    for i in d {
        if i == 1 { cnt += 1 }
        else { ret = max(ret, cnt); cnt = 0 }
    }
    
    print(ret)
}

main()